File: MemoryName.js
let currentUser = null;

// =====================
// 初期化
// =====================
window.onload = async () => {
  const res = await fetch("/api/me");
  if (res.ok) {
    currentUser = await res.json();
    showMainScreen();
  }
};

// =====================
// ログイン・ログアウト
// =====================
function showRegister() {
  document.getElementById("login-form").style.display = "none";
  document.getElementById("register-form").style.display = "block";
  document.getElementById("login-error").textContent = "";
}

function showLogin() {
  document.getElementById("register-form").style.display = "none";
  document.getElementById("login-form").style.display = "block";
  document.getElementById("register-error").textContent = "";
}

async function register() {
  const username = document.getElementById("reg-username").value;
  const password = document.getElementById("reg-password").value;
  const password2 = document.getElementById("reg-password2").value;
  const errEl = document.getElementById("register-error");

  if (!username || !password) { errEl.textContent = "ユーザー名とパスワードを入力してください"; return; }
  if (password !== password2) { errEl.textContent = "パスワードが一致しません"; return; }

  const res = await fetch("/api/register", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username, password })
  });

  const data = await res.json();
  if (!res.ok) { errEl.textContent = data.error; return; }

  alert("登録しました！ログインしてください");
  showLogin();
}

async function login() {
  const username = document.getElementById("login-username").value;
  const password = document.getElementById("login-password").value;

  const res = await fetch("/api/login", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username, password })
  });

  const data = await res.json();
  if (!res.ok) {
    document.getElementById("login-error").textContent = data.error;
    return;
  }

  currentUser = data.user;
  showMainScreen();
}

async function logout() {
  await fetch("/api/logout", { method: "POST" });
  currentUser = null;
  document.getElementById("main-screen").style.display = "none";
  document.getElementById("login-screen").style.display = "flex";
  document.getElementById("login-password").value = "";
  document.getElementById("login-error").textContent = "";
}

// =====================
// 画面切り替え
// =====================
function showMainScreen() {
  document.getElementById("login-screen").style.display = "none";
  document.getElementById("main-screen").style.display = "block";

  const roleLabel = { user: "利用者", admin: "管理者", family: "家族" };
  document.getElementById("header-user").textContent =
    `${currentUser.username}（${roleLabel[currentUser.role]}）`;

  buildTabs();
}

function buildTabs() {
  const tabBar = document.getElementById("tab-bar");
  tabBar.innerHTML = "";

  const tabs = getTabs();
  tabs.forEach((t, i) => {
    const btn = document.createElement("button");
    btn.className = "tab-btn" + (i === 0 ? " active" : "");
    btn.textContent = t.label;
    btn.onclick = () => switchTab(t.id);
    tabBar.appendChild(btn);
  });

  switchTab(tabs[0].id);
}

function getTabs() {
  const role = currentUser.role;
  if (role === 'admin') {
    return [
      { id: "admin-register", label: "📷 写真登録" },
      { id: "list",           label: "📋 登録一覧" },
      { id: "users",          label: "👤 ユーザー管理" },
      { id: "records",        label: "📊 記録" },
    ];
  } else if (role === 'family') {
    return [
      { id: "register", label: "📷 写真登録" },
      { id: "list",     label: "📋 登録一覧" },
    ];
  } else {
    return [
      { id: "game",     label: "🎮 ゲーム" },
      { id: "register", label: "📷 写真登録" },
      { id: "list",     label: "📋 登録一覧" },
      { id: "records",  label: "📊 記録" },
    ];
  }
}

function switchTab(tabId) {
  document.querySelectorAll(".tab-panel").forEach(p => p.classList.remove("active"));
  document.querySelectorAll(".tab-btn").forEach(b => b.classList.remove("active"));

  document.getElementById("tab-" + tabId).classList.add("active");

  const tabs = getTabs();
  const idx = tabs.findIndex(t => t.id === tabId);
  document.querySelectorAll(".tab-btn")[idx]?.classList.add("active");

  if (tabId === "list") {
    if (currentUser.role === 'admin') {
      document.getElementById("admin-user-select").style.display = "block";
      loadUserSelectOptions("list-user-select", true);
    }
    loadPeople();
  }
  if (tabId === "admin-register") loadUserSelectOptions("admin-owner-select", false);
  if (tabId === "users") { loadUsers(); loadUserSelectOptions("new-linked-user", false); }
  if (tabId === "records") loadRecords();
}

// =====================
// 写真管理
// =====================
async function addPerson() {
  const name = document.getElementById("person-name").value;
  const image = document.getElementById("person-image").files[0];
  if (!name || !image) { alert("名前と画像を入力してください"); return; }

  const formData = new FormData();
  formData.append("name", name);
  formData.append("image", image);

  const res = await fetch("/api/add_person", { method: "POST", body: formData });
  const data = await res.json();
  alert(res.ok ? "追加しました" : data.error);
  if (res.ok) {
    document.getElementById("person-name").value = "";
    document.getElementById("person-image").value = "";
    document.getElementById("file-drop-label").textContent = "📁 タップして写真を選ぶ";
  }
}

async function adminAddPerson() {
  const name = document.getElementById("admin-person-name").value;
  const image = document.getElementById("admin-person-image").files[0];
  const owner = document.getElementById("admin-owner-select").value;
  if (!name || !image || !owner) { alert("全て入力してください"); return; }

  const formData = new FormData();
  formData.append("name", name);
  formData.append("image", image);
  formData.append("owner_user_id", owner);

  const res = await fetch("/api/add_person", { method: "POST", body: formData });
  const data = await res.json();
  alert(res.ok ? "追加しました" : data.error);
  if (res.ok) {
    document.getElementById("admin-person-name").value = "";
    document.getElementById("admin-person-image").value = "";
    document.getElementById("admin-file-drop-label").textContent = "📁 タップして写真を選ぶ";
  }
}

async function loadPeople() {
  let url = "/api/get_people";
  if (currentUser.role === 'admin') {
    const sel = document.getElementById("list-user-select");
    if (sel && sel.value) url += `?owner_user_id=${sel.value}`;
  }

  const res = await fetch(url);
  const people = await res.json();
  const list = document.getElementById("registered-list");
  list.innerHTML = "";

  if (people.length === 0) {
    list.innerHTML = "<p class='empty-text'>まだ登録されていません</p>";
    return;
  }

  people.forEach(p => {
    const div = document.createElement("div");
    div.className = "person-card";
    div.innerHTML = `
      <img src="/api/get_image/${p.id}" alt="${p.name}">
      <p>${p.name}</p>
      <button class="delete-btn" onclick="deletePerson(${p.id})">削除</button>
    `;
    list.appendChild(div);
  });
}

async function deletePerson(id) {
  if (!confirm("削除しますか？")) return;
  const res = await fetch(`/api/delete_person/${id}`, { method: "DELETE" });
  const data = await res.json();
  alert(res.ok ? "削除しました" : data.error);
  if (res.ok) loadPeople();
}

// =====================
// ユーザー管理（管理者）
// =====================
async function loadUsers() {
  const res = await fetch("/api/users");
  const users = await res.json();
  const list = document.getElementById("user-list");
  list.innerHTML = "";

  const roleLabel = { user: "利用者", admin: "管理者", family: "家族" };
  const roleClass = { user: "role-user", admin: "role-admin", family: "role-family" };

  users.forEach(u => {
    const div = document.createElement("div");
    div.className = "user-row";
    div.innerHTML = `
      <div>
        <span class="user-name">${u.username}</span>
        <span class="role-badge ${roleClass[u.role]}">${roleLabel[u.role]}</span>
        ${u.linked_user_id ? `<span style="font-size:12px; color:#999; margin-left:4px;">紐づけID:${u.linked_user_id}</span>` : ""}
      </div>
      <button class="delete-btn" onclick="deleteUser(${u.id})">削除</button>
    `;
    list.appendChild(div);
  });
}

async function createUser() {
  const username = document.getElementById("new-username").value;
  const password = document.getElementById("new-password").value;
  const role = document.getElementById("new-role").value;
  const linked = document.getElementById("new-linked-user").value;

  const body = { username, password, role };
  if (role === 'family' && linked) body.linked_user_id = parseInt(linked);

  const res = await fetch("/api/users", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body)
  });
  const data = await res.json();
  alert(res.ok ? "ユーザーを作成しました" : data.error);
  if (res.ok) {
    document.getElementById("new-username").value = "";
    document.getElementById("new-password").value = "";
    loadUsers();
  }
}

async function deleteUser(id) {
  if (!confirm("このユーザーを削除しますか？")) return;
  const res = await fetch(`/api/users/${id}`, { method: "DELETE" });
  const data = await res.json();
  alert(res.ok ? "削除しました" : data.error);
  if (res.ok) loadUsers();
}

function toggleLinkedUser() {
  const role = document.getElementById("new-role").value;
  document.getElementById("linked-user-group").style.display =
    role === 'family' ? "block" : "none";
}

async function loadUserSelectOptions(selectId, includeAll) {
  const res = await fetch("/api/users");
  if (!res.ok) return;
  const users = await res.json();
  const sel = document.getElementById(selectId);
  if (!sel) return;

  const current = sel.value;
  sel.innerHTML = includeAll ? '<option value="">全員</option>' : "";

  users.filter(u => u.role === 'user').forEach(u => {
    const opt = document.createElement("option");
    opt.value = u.id;
    opt.textContent = u.username;
    sel.appendChild(opt);
  });

  if (current) sel.value = current;
}

// =====================
// ゲーム
// =====================
let gameTargets = [];
let answerPerson = null;

async function startMemoryGame() {
  const res = await fetch("/api/get_people");
  const people = await res.json();

  if (people.length < 2) {
    alert("最低2人登録してください");
    return;
  }

  const shuffled = [...people].sort(() => Math.random() - 0.5);
  gameTargets = shuffled.slice(0, 2);
  answerPerson = gameTargets[Math.floor(Math.random() * 2)];

  // すべてのゲームフェーズを隠す
  hideAllPhases();
  showMemorizePhase();
}

function hideAllPhases() {
  ['game-start', 'game-memorize', 'game-answer', 'game-correct', 'game-wrong'].forEach(id => {
    document.getElementById(id).style.display = 'none';
  });
}

function showMemorizePhase() {
  const grid = document.getElementById("memorize-grid");
  grid.innerHTML = "";

  gameTargets.forEach(p => {
    const div = document.createElement("div");
    div.className = "memory-card";
    div.innerHTML = `
      <img src="/api/get_image/${p.id}" alt="${p.name}">
      <p class="card-person-name">${p.name}</p>
    `;
    grid.appendChild(div);
  });

  document.getElementById("game-memorize").style.display = "block";

  let sec = 5;
  const countEl = document.getElementById("memory-countdown");
  countEl.textContent = sec;

  const timer = setInterval(() => {
    sec--;
    if (sec > 0) {
      countEl.textContent = sec;
    } else {
      clearInterval(timer);
      countEl.textContent = "";
      document.getElementById("game-memorize").style.display = "none";
      showAnswerPhase();
    }
  }, 1000);
}

function showAnswerPhase() {
  document.getElementById("answer-question").textContent = `「${answerPerson.name}」はどっち？`;
  const grid = document.getElementById("answer-grid");
  grid.innerHTML = "";

  [...gameTargets].sort(() => Math.random() - 0.5).forEach(p => {
    const div = document.createElement("div");
    div.className = "answer-card";
    div.innerHTML = `<img src="/api/get_image/${p.id}" alt="">`;
    div.onclick = () => checkAnswer(p);
    grid.appendChild(div);
  });

  document.getElementById("game-answer").style.display = "block";
}

async function checkAnswer(chosen) {
  // クリック無効化
  document.querySelectorAll(".answer-card").forEach(c => c.onclick = null);

  const isCorrect = chosen.id === answerPerson.id;

  // 記録保存
  await fetch("/api/save_record", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ correct: isCorrect ? 1 : 0, total: 1 })
  });

  hideAllPhases();

  if (isCorrect) {
    // 正解画面
    document.getElementById("correct-photo").src = `/api/get_image/${answerPerson.id}`;
    document.getElementById("correct-photo").alt = answerPerson.name;
    document.getElementById("correct-name").textContent = answerPerson.name;
    document.getElementById("correct-photo-label").textContent = `「${answerPerson.name}」`;
    document.getElementById("correct-detail").textContent = `正解！「${answerPerson.name}」さんを選びました。`;
    document.getElementById("game-correct").style.display = "block";
  } else {
    // 不正解画面
    document.getElementById("wrong-chosen-photo").src = `/api/get_image/${chosen.id}`;
    document.getElementById("wrong-chosen-name").textContent = chosen.name || "？";

    const correct = gameTargets.find(p => p.id === answerPerson.id);
    document.getElementById("wrong-correct-photo").src = `/api/get_image/${answerPerson.id}`;
    document.getElementById("wrong-correct-name").textContent = answerPerson.name;

    document.getElementById("game-wrong").style.display = "block";
  }
}

// =====================
// 記録
// =====================
async function loadRecords() {
  const res = await fetch("/api/get_records");
  const records = await res.json();
  const list = document.getElementById("records-list");

  if (records.length === 0) {
    list.innerHTML = "<p class='empty-text'>記録がありません</p>";
    return;
  }

  let html = `<table class="records-table">
    <thead>
      <tr>
        ${currentUser.role === 'admin' ? '<th>ユーザー</th>' : ""}
        <th>結果</th>
        <th>日時</th>
      </tr>
    </thead><tbody>`;

  records.forEach(r => {
    const date = new Date(r.played_at).toLocaleString('ja-JP');
    html += `<tr>
      ${currentUser.role === 'admin' ? `<td>${r.username}</td>` : ""}
      <td class="${r.correct ? 'record-correct' : 'record-wrong'}">
        ${r.correct ? "正解 ✓" : "不正解 ✗"}
      </td>
      <td class="record-date">${date}</td>
    </tr>`;
  });

  html += "</tbody></table>";
  list.innerHTML = html;
}

File: MemoryName.py
from flask import Flask, request, jsonify, send_from_directory, session
import os
from functools import wraps
import psycopg2
from psycopg2.extras import RealDictCursor

app = Flask(__name__)
app.secret_key = os.environ.get("SECRET_KEY", "memoryname-secret-key-2024")

DATABASE_URL = os.environ.get("DATABASE_URL")

# --- DB接続 ---
def get_db():
    conn = psycopg2.connect(DATABASE_URL, cursor_factory=RealDictCursor)
    return conn

# --- DB初期化 ---
def init_db():
    conn = get_db()
    c = conn.cursor()

    c.execute("""
    CREATE TABLE IF NOT EXISTS users (
        id SERIAL PRIMARY KEY,
        username TEXT UNIQUE NOT NULL,
        password TEXT NOT NULL,
        role TEXT NOT NULL CHECK(role IN ('user', 'admin', 'family')),
        linked_user_id INTEGER
    )
    """)

    c.execute("""
    CREATE TABLE IF NOT EXISTS people (
        id SERIAL PRIMARY KEY,
        name TEXT,
        image BYTEA,
        owner_user_id INTEGER NOT NULL
    )
    """)

    c.execute("""
    CREATE TABLE IF NOT EXISTS game_records (
        id SERIAL PRIMARY KEY,
        player_user_id INTEGER NOT NULL,
        correct INTEGER NOT NULL,
        total INTEGER NOT NULL,
        played_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
    )
    """)

    c.execute("SELECT id FROM users WHERE username = 'admin'")
    if not c.fetchone():
        c.execute("INSERT INTO users (username, password, role) VALUES (%s, %s, %s)",
                  ('admin', 'admin123', 'admin'))

    conn.commit()
    conn.close()
    print("DB初期化完了")

# --- 認証デコレーター ---
def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'user_id' not in session:
            return jsonify({"error": "ログインが必要です"}), 401
        return f(*args, **kwargs)
    return decorated

def role_required(*roles):
    def decorator(f):
        @wraps(f)
        def decorated(*args, **kwargs):
            if 'user_id' not in session:
                return jsonify({"error": "ログインが必要です"}), 401
            if session.get('role') not in roles:
                return jsonify({"error": "権限がありません"}), 403
            return f(*args, **kwargs)
        return decorated
    return decorator

# --- ページ配信 ---
@app.route("/")
def home():
    return send_from_directory(".", "index.html")

@app.route("/game")
def game():
    return send_from_directory(".", "index.html")

@app.route("/<path:filename>")
def static_files(filename):
    return send_from_directory(".", filename)

# =====================
# 認証API
# =====================

@app.route("/api/register", methods=["POST"])
def register():
    data = request.get_json()
    username = data.get("username")
    password = data.get("password")

    if not username or not password:
        return jsonify({"error": "入力が不正です"}), 400

    conn = get_db()
    c = conn.cursor()
    try:
        c.execute("INSERT INTO users (username, password, role) VALUES (%s, %s, %s)",
                  (username, password, "user"))
        conn.commit()
    except Exception:
        conn.rollback()
        conn.close()
        return jsonify({"error": "そのユーザー名は既に使われています"}), 400
    conn.close()
    return jsonify({"message": "登録しました"})

@app.route("/api/login", methods=["POST"])
def login():
    data = request.get_json()
    username = data.get("username")
    password = data.get("password")

    conn = get_db()
    c = conn.cursor()
    c.execute("SELECT * FROM users WHERE username = %s AND password = %s", (username, password))
    user = c.fetchone()
    conn.close()

    if not user:
        return jsonify({"error": "ユーザー名またはパスワードが違います"}), 401

    session['user_id'] = user['id']
    session['username'] = user['username']
    session['role'] = user['role']
    session['linked_user_id'] = user['linked_user_id']

    return jsonify({
        "message": "ログイン成功",
        "user": {
            "id": user['id'],
            "username": user['username'],
            "role": user['role'],
            "linked_user_id": user['linked_user_id']
        }
    })

@app.route("/api/logout", methods=["POST"])
def logout():
    session.clear()
    return jsonify({"message": "ログアウトしました"})

@app.route("/api/me", methods=["GET"])
def me():
    if 'user_id' not in session:
        return jsonify({"error": "未ログイン"}), 401
    return jsonify({
        "id": session['user_id'],
        "username": session['username'],
        "role": session['role'],
        "linked_user_id": session['linked_user_id']
    })

# =====================
# ユーザー管理API
# =====================

@app.route("/api/users", methods=["GET"])
@role_required('admin')
def get_users():
    conn = get_db()
    c = conn.cursor()
    c.execute("SELECT id, username, role, linked_user_id FROM users")
    rows = c.fetchall()
    conn.close()
    return jsonify([dict(r) for r in rows])

@app.route("/api/users", methods=["POST"])
@role_required('admin')
def create_user():
    data = request.get_json()
    username = data.get("username")
    password = data.get("password")
    role = data.get("role")
    linked_user_id = data.get("linked_user_id")

    if not username or not password or role not in ('user', 'admin', 'family'):
        return jsonify({"error": "入力が不正です"}), 400

    conn = get_db()
    c = conn.cursor()
    try:
        c.execute("INSERT INTO users (username, password, role, linked_user_id) VALUES (%s, %s, %s, %s)",
                  (username, password, role, linked_user_id))
        conn.commit()
    except Exception:
        conn.rollback()
        conn.close()
        return jsonify({"error": "そのユーザー名は既に使われています"}), 400
    conn.close()
    return jsonify({"message": "ユーザーを作成しました"})

@app.route("/api/users/<int:user_id>", methods=["DELETE"])
@role_required('admin')
def delete_user(user_id):
    conn = get_db()
    c = conn.cursor()
    c.execute("DELETE FROM users WHERE id = %s", (user_id,))
    conn.commit()
    conn.close()
    return jsonify({"message": "削除しました"})

# =====================
# 写真管理API
# =====================

def get_target_user_id():
    role = session.get('role')
    if role == 'family':
        return session.get('linked_user_id')
    return session.get('user_id')

@app.route("/api/add_person", methods=["POST"])
@login_required
def add_person():
    role = session.get('role')
    name = request.form.get("name")
    image = request.files.get("image")

    if not name or not image:
        return jsonify({"error": "name and image required"}), 400

    if role == 'admin':
        owner_id = request.form.get("owner_user_id", session['user_id'])
    else:
        owner_id = get_target_user_id()

    if not owner_id:
        return jsonify({"error": "紐づけ先ユーザーが設定されていません"}), 400

    image_data = image.read()
    conn = get_db()
    c = conn.cursor()
    c.execute("INSERT INTO people (name, image, owner_user_id) VALUES (%s, %s, %s)",
              (name, psycopg2.Binary(image_data), owner_id))
    conn.commit()
    conn.close()
    return jsonify({"message": "saved"})

@app.route("/api/get_people", methods=["GET"])
@login_required
def get_people():
    role = session.get('role')
    conn = get_db()
    c = conn.cursor()

    if role == 'admin':
        target_id = request.args.get("owner_user_id")
        if target_id:
            c.execute("SELECT id, name, owner_user_id FROM people WHERE owner_user_id = %s", (target_id,))
        else:
            c.execute("SELECT id, name, owner_user_id FROM people")
    else:
        owner_id = get_target_user_id()
        c.execute("SELECT id, name, owner_user_id FROM people WHERE owner_user_id = %s", (owner_id,))

    rows = c.fetchall()
    conn.close()
    return jsonify([dict(r) for r in rows])

@app.route("/api/get_image/<int:person_id>", methods=["GET"])
@login_required
def get_image(person_id):
    conn = get_db()
    c = conn.cursor()
    c.execute("SELECT image FROM people WHERE id = %s", (person_id,))
    row = c.fetchone()
    conn.close()

    if row is None:
        return "Not found", 404

    return bytes(row["image"]), 200, {"Content-Type": "image/jpeg"}

@app.route("/api/delete_person/<int:person_id>", methods=["DELETE"])
@login_required
def delete_person(person_id):
    role = session.get('role')
    conn = get_db()
    c = conn.cursor()
    c.execute("SELECT owner_user_id FROM people WHERE id = %s", (person_id,))
    row = c.fetchone()

    if row is None:
        conn.close()
        return jsonify({"error": "見つかりません"}), 404

    if role != 'admin' and row['owner_user_id'] != get_target_user_id():
        conn.close()
        return jsonify({"error": "権限がありません"}), 403

    c.execute("DELETE FROM people WHERE id = %s", (person_id,))
    conn.commit()
    conn.close()
    return jsonify({"message": "deleted"})

# =====================
# 記録API
# =====================

@app.route("/api/save_record", methods=["POST"])
@login_required
def save_record():
    data = request.get_json()
    correct = data.get("correct", 0)
    total = data.get("total", 0)
    player_id = session['user_id']

    conn = get_db()
    c = conn.cursor()
    c.execute("INSERT INTO game_records (player_user_id, correct, total) VALUES (%s, %s, %s)",
              (player_id, correct, total))
    conn.commit()
    conn.close()
    return jsonify({"message": "記録しました"})

@app.route("/api/get_records", methods=["GET"])
@login_required
def get_records():
    role = session.get('role')
    conn = get_db()
    c = conn.cursor()

    if role == 'admin':
        c.execute("""
            SELECT r.id, u.username, r.correct, r.total, r.played_at
            FROM game_records r JOIN users u ON r.player_user_id = u.id
            ORDER BY r.played_at DESC LIMIT 50
        """)
    else:
        c.execute("""
            SELECT r.id, u.username, r.correct, r.total, r.played_at
            FROM game_records r JOIN users u ON r.player_user_id = u.id
            WHERE r.player_user_id = %s
            ORDER BY r.played_at DESC LIMIT 50
        """, (session['user_id'],))

    rows = c.fetchall()
    conn.close()
    return jsonify([dict(r) for r in rows])

# gunicorn用DB初期化
with app.app_context():
    init_db()

if __name__ == "__main__":
    init_db()
    app.run(host="0.0.0.0", port=5000, debug=True)
